File: Codigos estudiantes por lenguaje/Java/Daniel Cangrejo/Ordenamientos/Tim_Sort.java
import java.util.Scanner;

public class Main{
	
	static int RUN=32;
	
	static void insertionSort(int arr[], int left, int right) {
	for (int i = left+1; i <= right; i++) {
		int temp = arr[i];
		int j=i-1;
		while(arr[j] > temp && j>=left) {
			arr[j+1]=arr[j];
			j--;
			if (j<0) {
				break;
			}
		}
		arr[j+1]=temp;
	 }
	}

	static void merge(int arr[],int left, int mid, int right) {
		int len1 = mid-left+1, len2 = right-mid;
		int L[]=new int [len1];
		int R[]= new int [len2];
		for (int x = 0; x < len1; x++) {
			L[x]=arr[left+x];
		}
		for (int x = 0; x < len2; x++) {
			R[x]=arr[mid+1+x];
		}
		int i=0,j=0,k=left;
		while (i<len1 && j<len2) {
			if (L[i]<=R[j]) {
				arr[k]=L[i];
				i++;
			}else {
				arr[k]=R[j];
				j++;
			}
			k++;
			
		}
		while(i<len1) {
			arr[k]=L[i];
			i++;
			k++;
		}
		while(j<len2) {
			arr[k]=R[j];
			j++;
			k++;
		}
	}	
	
	static void timSort(int arr[], int n) {
		for (int i = 0; i < n; i+=RUN) {
			insertionSort(arr, i, Math.min((i+31),(n-1)));
			
		}
		for (int size = RUN; size < n; size=2*size) {
			for (int left = 0; left < n; left+=2*size) {
				int mid=left+size-1;
				int right = Math.min((left+2*size-1), (n-1));
				merge(arr, left, mid, right);
			}
		}
	}
	
	public static void printArray(int[] arr, int n) {
		for (int i = 0; i < n; i++) {
			System.out.print(arr[i]+" ");
		}
		System.out.println();
	}

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int arr[]= {5,21,7,23,19};
        int n=arr.length;
        printArray(arr, n);
        timSort(arr, n);
        printArray(arr, n);
    }
}

File: Codigos estudiantes por lenguaje/PY/Daniel Cangrejo/Ordenamientos/Tim_Sort.py
from sys import stdin, stdout
from random import randint
rl = stdin.readline
wr = stdout.write


RUN = 32


def insertionSort(arr, left, right):

    for i in range(left+1, right+1):
        temp = arr[i]
        j = i - 1
        while arr[j] > temp and j >= left:
            arr[j+1] = arr[j]
            j -= 1
            if j < 0:
                break

        arr[j+1] = temp


def mergeSort(arr, left, mid, right):

    len1 = mid-left+1
    len2 = right-mid

    L = [0] * len1
    R = [0] * len2

    for x in range(len1):
        L[x] = arr[left+x]
    for x in range(len2):
        R[x] = arr[mid+1+x]

    i, j, k = 0, 0, left
    while i < len1 and j < len2:
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1

    while i < len1:
        arr[k] = L[i]
        i += 1
        k += 1
    while j < len2:
        arr[k] = R[j]
        j += 1
        k += 1


def timSort(arr, n):

    i = 0
    while i < n:
        insertionSort(arr, i, min((i+31), (n-1)))
        size = RUN
        while size < n:
            left = 0
            while left < n:
                mid = min((left + size - 1), (n-1))
                right = min((left+2*size-1), (n-1))
                mergeSort(arr, left, mid, right)
                left += 2 * size
            size *= 2
        i += RUN


def printArray(arr, n):

    for i in arr:
        wr(f'{i} ')
    wr('\n')


arr = list(map(int, rl().split()))
n = len(arr)
printArray(arr, n)
timSort(arr, n)
printArray(arr, n)
